packages/fxa-auth-server/lib/routes/totp.js

```diff
@@ -117,69 +101,52 @@ module.exports = (log, db, mailer, customs, config) => {
-          // before TOTP was enabled, can remove TOTP. Any new sessions after TOTP is enabled, are only considered
-          // verified *if and only if* they have verified a TOTP code.
-          if (!sessionToken.tokenVerified) {
-            throw errors.unverifiedSession();
-          }
+        // If a TOTP token is not verified, we should be able to safely delete regardless of session
+        // verification state.
+        const hasEnabledToken = await totpUtils.hasTotpToken({ uid });
 
-          return db.deleteTotpToken(uid).then(() => {
-            return log.notifyAttachedServices('profileDataChanged', request, {
-              uid: sessionToken.uid,
-            });
-          });
+        // To help prevent users from getting locked out of their account, sessions created and verified
+        // before TOTP was enabled, can remove TOTP. Any new sessions after TOTP is enabled, are only considered
+        // verified *if and only if* they have verified a TOTP code.
+        if (!sessionToken.tokenVerified) {
+          throw errors.unverifiedSession();
         }
 
-        function sendEmailNotification() {
-          if (!hasEnabledToken) {
-            return;
-          }
+        await db.deleteTotpToken(uid);
 
-          return db.account(sessionToken.uid).then(account => {
-            const geoData = request.app.geo;
-            const ip = request.app.clientAddress;
-            const emailOptions = {
-              acceptLanguage: request.app.acceptLanguage,
-              ip: ip,
-              location: geoData.location,
-              timeZone: geoData.timeZone,
-              uaBrowser: request.app.ua.browser,
-              uaBrowserVersion: request.app.ua.browserVersion,
-              uaOS: request.app.ua.os,
-              uaOSVersion: request.app.ua.osVersion,
-              uaDeviceType: request.app.ua.deviceType,
-              uid: sessionToken.uid,
-            };
-
-            mailer.sendPostRemoveTwoStepAuthNotification(
-              account.emails,
-              account,
-              emailOptions
-            );
-          });
+        await log.notifyAttachedServices('profileDataChanged', request, {
+          uid,
+        });
+
+        if (hasEnabledToken) {
+          const account = await db.account(uid);
+          const geoData = request.app.geo;
+          const ip = request.app.clientAddress;
+          const emailOptions = {
+            acceptLanguage: request.app.acceptLanguage,
+            ip,
+            location: geoData.location,
+            timeZone: geoData.timeZone,
+            uaBrowser: request.app.ua.browser,
+            uaBrowserVersion: request.app.ua.browserVersion,
+            uaOS: request.app.ua.os,
+            uaOSVersion: request.app.ua.osVersion,
+            uaDeviceType: request.app.ua.deviceType,
+            uid,
+          };
+
+          await mailer.sendPostRemoveTwoStepAuthNotification(
```

Comment:
Would be good to have @rfk verify this should be waited on here, as previously the promise was not returned to be waited on.

Reply:
Hrm; to be honest I don't recall. Most of the places where we deliberately don't await a promise have a comment about why. I guess the only reason not to wait here would be to avoid a mail-sending error from "failing" the entire request, making it look like we failed to update the account when we really did. If we're worried about that, I think it would be better to explicitly await and then catch-and-ignore the error.